### ISSUE TITLE
feat: implement MLOps data feedback and retraining loop

### ISSUE THREAD

**Issue by sharp231** [sharp231/classifier#2]
### 概要
フィードバック機能による継続的なデータ収集と再学習ループ（MLOps）を構築する。

### 対象
- 収集データの品質管理
- 再学習トリガー条件の検討
- 保存/参照フローの設計

**Comment by sharp231** [comment]
## 概要

ユーザーから画像判定結果へのフィードバックを収集し、データ品質の確認、モデル評価、再学習、モデル更新までを安全に実行できるMLOpsループを段階的に構築する。

このIssueは複数の子Issueをまとめる上位Issueとして扱い、一度のPull Requestですべてを実装しない。

## 背景

現在のアプリケーションでは、画像を分類して結果を表示できるが、判定結果が正しかったかを収集する仕組みがない。

モデルを継続的に改善するには、次の流れが必要になる。

```text
画像判定
  ↓
ユーザーフィードバック
  ↓
データ検証
  ↓
人による確認
  ↓
評価データ・学習データへの採用
  ↓
モデル再学習
  ↓
旧モデルとの比較
  ↓
承認
  ↓
段階的なモデル更新
  ↓
問題発生時のロールバック
```

ただし、ユーザーから得たフィードバックをそのまま正解データとして扱うことはできない。

## 目的

- 実際の利用データからモデル改善の候補を把握する
- 誤分類が多いクラスや入力傾向を分析する
- モデルとデータのバージョンを追跡できるようにする
- 再学習前後の性能を比較できるようにする
- 性能が悪化したモデルを本番へ公開しない
- 問題が発生した場合に旧モデルへ戻せるようにする
- ユーザー画像を必要以上に収集しない

## 前提条件

このIssueへ着手する前に、以下を完了または決定する。

- フィードバックデータの利用目的
- 保存するデータ項目
- 画像を保存するか
- ユーザー同意の方法
- 保存期間
- 削除方法
- モデルバージョンの管理方法
- 前処理バージョンの管理方法
- TTAの評価結果
- 現行モデルの基準性能
- 新旧モデルの採用基準

前提が確定するまで `status:hold` とする。

## フェーズ

### Phase 1: フィードバック仕様の定義

次を決定する。

- フィードバックの種類
- 保存項目
- 保存しない項目
- 画像保存の有無
- 同意方法
- 保存期間
- 削除方法
- 不正送信への対策
- データ品質の確認方法

成果物候補:

```text
docs/feedback-data-policy.md
```

### Phase 2: フィードバックUI

結果画面に、判定結果へのフィードバック操作を追加する。

例:

```text
判定結果: 猫

[正しい]
[間違っている]
[この画像は対象外]
```

「間違っている」を選択した場合に正解ラベルを選択させるかを決定する。

フィードバック送信の失敗、二重送信、再送信を考慮する。

### Phase 3: フィードバックAPI

フィードバックを受け付けるAPIを追加する。

API候補:

```text
POST /api/v1/predictions/{prediction_id}/feedback
```

入力例:

```json
{
  "feedback_type": "incorrect",
  "correct_label": "犬"
}
```

APIでは次を検証する。

- prediction IDの形式
- 対象predictionの存在
- 許可されたfeedback type
- correct labelの許可値
- 同一predictionへの重複送信
- 更新可能な項目
- リクエストサイズ
- rate limitの必要性
- 権限と所有関係

### Phase 4: データ保存

保存対象のデータモデルを定義する。

候補:

```text
predictions
├── id
├── model_version
├── preprocessing_version
├── predicted_label
├── confidence_score
├── top_k_scores
└── created_at

feedback
├── id
├── prediction_id
├── feedback_type
├── correct_label
├── consent_version
├── created_at
└── reviewed_at
```

画像を保存する場合は、DBへ画像本体を直接保存するのではなく、オブジェクトストレージと参照情報を分離することを検討する。

### Phase 5: データ品質確認

収集したデータを自動的に学習へ渡さず、次を確認する。

- 重複データ
- 破損データ
- 不正ラベル
- CIFAR-10対象外の画像
- 複数物体を含む画像
- 矛盾したフィードバック
- 意図的な誤ラベル
- 自動化されたスパム
- クラス分布の偏り
- モデルバージョンの違い
- 前処理バージョンの違い

学習データへ採用する前に、人による確認を挟む。

### Phase 6: 再学習

最初は手動実行可能な再学習処理を作成する。

再学習処理では次を固定・記録する。

- 学習データのバージョン
- 評価データのバージョン
- ランダムシード
- ハイパーパラメータ
- Pythonバージョン
- TensorFlow/Kerasバージョン
- 前処理バージョン
- 出力モデルの識別子

定期実行や自動再学習は、手動フローが安定してから検討する。

### Phase 7: モデル評価

新モデルを本番へ反映する前に、現行モデルと比較する。

最低限、次を評価する。

- Top-1 Accuracy
- Top-3 Accuracy
- Log Loss
- クラス別Accuracy
- confidence calibration
- 推論時間
- モデル読み込み時間
- メモリ使用量
- モデルファイルサイズ
- 判定困難機能への影響

全体精度が向上していても、特定クラスが大幅に悪化していないか確認する。

### Phase 8: モデル承認と公開

評価を通過したモデルだけを本番候補とする。

次を記録する。

- モデルバージョン
- 学習データバージョン
- 評価結果
- 承認者
- 公開日時
- 対象環境
- 以前のモデルバージョン
- ロールバック方法

収集データを受け取っただけで、自動的に再学習・本番公開しない。

## セキュリティとプライバシー

- 画像保存は明示的な同意を必要とする
- 画像提供をデフォルトで有効にしない
- 必要最小限のデータだけを保存する
- 保存期間を設定する
- 保存期限後に削除する
- 削除要求へ対応できる構造にする
- prediction IDだけで他者のデータを操作できないようにする
- ユーザー指定のモデルバージョンや正解データを信用しない
- DB更新列をallowlistで制限する
- SQLへ外部入力を直接連結しない
- 画像や個人情報を通常ログへ記録しない
- 管理用操作とユーザー用操作を分離する
- 再学習処理から本番Secretsへ不要にアクセスさせない

## テスト方針

### フロントエンド

- 正常にフィードバックを送信できる
- 送信中状態を表示できる
- 送信成功を表示できる
- 送信失敗後に再試行できる
- 二重送信を防止できる

### API

- 正常なフィードバックを受け付ける
- 不正なfeedback typeを拒否する
- 不正なcorrect labelを拒否する
- 存在しないpredictionを処理する
- 重複送信を安全に処理する
- 許可されていない項目を更新しない
- 過大なリクエストを拒否する

### DB

- predictionとfeedbackの関係を保つ
- 存在しないpredictionへのfeedbackを拒否する
- 重複制約を確認する
- トランザクション失敗時に部分更新を残さない
- 削除方針を確認する
- マイグレーションを検証する

### 再学習

- 同じ入力と設定から再現可能な結果を生成する
- 不正データを学習対象へ含めない
- 評価基準を満たさないモデルを不採用にする
- 現行モデルを上書きせず、新しいバージョンとして保存する
- ロールバック対象のモデルを保持する

## スコープ外

初期段階では以下を行わない。

- ユーザー同意なしの画像収集
- フィードバックを即座に正解データとして採用すること
- 無審査の自動再学習
- 無審査の自動本番デプロイ
- Airflow、DVC、MLflowなどの一括導入
- 複数のDB製品の同時導入
- 本番データの破壊的マイグレーション
- YOLOへの移行
- リアルタイム学習
- A/Bテストの同時実装

## 完了条件

この上位Issueは、少なくとも次を満たしたときに完了とする。

- フィードバックデータ方針が文書化されている
- ユーザー同意と画像保存方針が定義されている
- フィードバックUIとAPIが実装されている
- DB制約とマイグレーションがテストされている
- データ品質確認フローが存在する
- 再学習処理が再現可能である
- 新旧モデルの評価結果を比較できる
- 評価基準を満たさないモデルを公開しない
- モデルバージョンを追跡できる
- 旧モデルへロールバックできる
- 関連テストと全テストが成功する
- 保存期間と削除処理が実装されている

## 子Issue候補

- `docs: フィードバックデータの保存方針を定義する`
- `feat: 判定結果へのフィードバックUIを追加する`
- `feat: フィードバック受付APIを追加する`
- `feat: フィードバック保存用DBスキーマを追加する`
- `test: フィードバックデータの品質検証を追加する`
- `feat: 再学習スクリプトを再現可能にする`
- `test: 新旧モデルの評価ゲートを追加する`
- `feat: モデルバージョン管理を追加する`
- `docs: モデル公開とロールバック手順を定義する`

## 依存関係

- フィードバックデータの保存方針
- 通常推論とTTAの比較
- モデルバージョンの定義
- ヘルスチェックと推論時間計測
